src/sync/subscription_state.hpp

```diff
@@ -0,0 +1,37 @@
+////////////////////////////////////////////////////////////////////////////
+//
+// Copyright 2018 Realm Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+////////////////////////////////////////////////////////////////////////////
+
+#ifndef REALM_OS_SUBSCRIPTION_STATE_HPP
+#define REALM_OS_SUBSCRIPTION_STATE_HPP
+
+#include <cstdint>
+
+namespace realm {
+namespace partial_sync {
+
+// Enum describing the various states a partial sync subscription can have.
+enum class SubscriptionState : int8_t {
+    Error = -1,             // An error occurred with the partial sync subscription..
+    Uninitialized = 0,      // The subscription was created, but has not yet been processed by the sync server.
+    Initialized = 1         // The subscription has been initialized successfully and is syncing data to the device.
```

Comment:
I'm thinking that `Incomplete`/ `Complete` or `Pending`/`Complete` may be better terms here, since "uninitialized" doesn't seem like a great term for a subscription that we've created but that the server simply hasn't processed yet. It's accurate in describing the `List` of results associated with the subscription, but not for the subscription itself.

On another note, is `State` or `Status` the tight term here? I used `Subscription::status()` because it felt slightly better to me, but it's obviously bad to be inconsistent within the same API.

Reply:
New/Pending/Complete?

Reply:
What would the distinction be between the new and pending states?

Reply:
I could see some merit in introducing a state that represents the period while we're waiting on the subscription to be validated and created by the background thread. "New" doesn't seem like a great fit for that though, since there's no guarantee that the background thread will actually create the subscription (it may fail one of the validation checks, or it may be an exact match for an existing subscription). Something like "Creating" may work, though.

Reply:
That was my thinking when I commented on https://github.com/realm/realm-object-store/pull/610#discussion_r164691985, although I may not have articulated that well enough.

I would even argue for adding a callback on the `subscribe` call that will be invoked when the subscription has been created and written to the background Realm. My reasoning (beyond making it symmetric to the way I imagine `unsubscribe` will behave from the public API POV) is that it will differentiate waiting for the server from waiting for the client and once your callback is invoked, you'll have durability guarantees about your subscription. On the other hand, it may be confusing that the callback is invoked when the subscription is created rather than when its data has been synchronized.

I don't have strong feelings about either approach (callback or additional state), but I do feel strongly we need to add a way to notify the user that subscription request is written and waiting to be processed.

Reply:
The reason I opted not to use a callback as you describe is that it results in there being two places that errors may be reported for the same subscription: the callback (for an error in creating the subscription), and the subscription state change handler (for an error reported by the server).

> I do feel strongly we need to add a way to notify the user that subscription request is written and waiting to be processed.

To be clear, you're saying that you feel it's important to distinguish between "waiting for the subscription to be created by the background thread" and "waiting for the server to process the new subscription"?

Reply:
Yes, if that's easy to do, I see some merit in exposing it to developers so there's an early point in time where they can be sure that the subscription is written to the realm - so they can e.g. request additional execution time by the system and ensure that next time the app starts, it'll start synchronizing the subscribed query. I believe that was a pattern available on iOS a while back, not sure if it still applies today, but I imagine there are other similar use cases.

Reply:
I'm not sure the distinction matters, but I don't have any strong feelings neither towards State vs. Status nor the individual names used.

Reply:
I added a `Creating` state, used until the background thread has finished writing to the Realm, and renamed the other states to `Pending` and `Complete`.